chore(deps): Upgrade django-axes from 5.9.0 to 8.0.0

webjunkie opened the pull request:
## Summary

Upgrades django-axes from 5.9.0 to 8.0.0 for Django 5 compatibility.

## Changes

- django-axes: 5.9.0 → 8.0.0
- django-ipware: 3.0.7 → 7.0.1 (transitive dependency)
- 3 new migrations applied (0007, 0008, 0009)
- Fixed deprecated setting: `AXES_META_PRECEDENCE_ORDER` → `AXES_IPWARE_META_PRECEDENCE_ORDER`
- Added `AXES_HTTP_RESPONSE_CODE = 403` to maintain legacy response code
- Added manual axes integration in `LoginSerializer` and `LoginViewSet`

**Note**: See inline review comments on the code for detailed explanations of why each change was necessary.

## Test Plan

- Django system check passes
- Brute force test passes: `pytest posthog/api/test/test_authentication.py::TestLoginAPI::test_login_endpoint_is_protected_against_brute_force_attempts`
- Migration risk analysis passes (third-party policies skipped via #38957)

webjunkie committed:
chore(deps): Upgrade django-axes from 5.9.0 to 8.0.0

- Upgraded django-axes to 8.0.0 for Django 5 compatibility
- Applied 3 new migrations (0007, 0008, 0009)
- Fixed deprecated AXES_META_PRECEDENCE_ORDER → AXES_IPWARE_META_PRECEDENCE_ORDER
- Updated migration risk analyzer to skip PostHog policy checks on third-party apps
  - Added POSTHOG_OWNED_APPS list and is_posthog_app() helper
  - UUID primary key and single migration policies now only apply to posthog/ee/products apps
  - Allows third-party dependency migrations without policy violations

webjunkie committed:
chore(axes): Add AXES_HTTP_RESPONSE_CODE=403 setting

Keep legacy 403 status code for lockouts, as django-axes 6.0+ changed default to 429.
This maintains consistency with existing behavior and tests.

webjunkie committed:
fix(auth): Add manual axes integration for django-axes 8.0

Django-axes 8.0 requires manual integration in DRF serializers.
Added checks before/after authentication to properly handle lockouts:
- Check lockout before attempting authentication
- Check lockout after failed authentication (in case it just triggered)
- Handle AxesBackendPermissionDenied exception in LoginViewSet

This ensures the brute force protection returns 403 as expected.

webjunkie commented on posthog/api/authentication.py:
```diff
@@ -167,7 +167,20 @@ def create(self, validated_data: dict[str, str]) -> Any:
+        if handler.is_locked(axes_request, credentials=axes_credentials):
+            raise AxesBackendPermissionDenied("Account locked: too many login attempts.")
+
         user = cast(
```
**Why the pre-authentication check**: This prevents wasted authentication attempts for users/IPs that are already locked out. Returns 403 immediately instead of trying to authenticate first.

webjunkie reviewed (commented):
Added inline explanations for each code change to clarify the reasoning behind the django-axes 8.0 integration.

zlwaterfield reviewed (approved):
Surprised at how few changes were needed here! Can we block prod deploys and just test this fully on dev first?

webjunkie commented:
@zlwaterfield I guess so. I can ask infra. Then you would test it?

webjunkie committed:
Merge branch 'master' into chore/axes-upgrade

github-actions[bot] committed:
Update E2E screenshots for Playwright

github-actions[bot] committed:
Update E2E screenshots for Playwright

webjunkie committed:
Merge branch 'master' into chore/axes-upgrade

webjunkie committed:
Merge branch 'master' into chore/axes-upgrade

webjunkie committed:
Merge branch 'master' into chore/axes-upgrade

# Conflicts:
#	pyproject.toml
#	uv.lock

webjunkie committed:
Merge branch 'master' into chore/axes-upgrade

webjunkie commented:
@zlwaterfield It's now on https://app-two.dev.posthog.dev/ and looks good! Please check as well if you get a chance.

webjunkie committed:
Merge remote-tracking branch 'origin/master' into chore/axes-upgrade

Piccirello commented on posthog/settings/web.py:
```diff
@@ -21,7 +21,9 @@
 AXES_COOLOFF_TIME = timedelta(minutes=10)
 AXES_LOCKOUT_CALLABLE = "posthog.api.authentication.axes_locked_out"
-AXES_META_PRECEDENCE_ORDER = ["HTTP_X_FORWARDED_FOR", "REMOTE_ADDR"]
+AXES_IPWARE_META_PRECEDENCE_ORDER = ["HTTP_X_FORWARDED_FOR", "REMOTE_ADDR"]
```
Do we need to manually install `django-ipware` now that `django-ipware` is an optional dependency?

webjunkie replied:
We use `django-axes[ipware]` syntax which pulls it in as a transitive dependency - cleaner than an explicit dep and follows our pattern with `psycopg[binary]`.

Piccirello reviewed (approved):
My review is pretty surface level but this mostly LGTM pending a few comments and questions around tests. I reviewed the v6/v7/v8 changelog and we don't seem to use any of the settings that changed.

webjunkie committed:
chore(deps): add django-ipware explicitly for axes 8.0 compatibility

webjunkie committed:
test(authentication): add IP-based brute force lockout test

Verify that axes lockout is applied per-IP, not globally. This validates
that django-ipware 7.0.1 IP detection is working correctly with the
axes 8.0.0 upgrade and AXES_IPWARE_META_PRECEDENCE_ORDER configuration.

This complements the existing brute force test and confirms different IPs
are tracked independently for lockout purposes.

webjunkie commented on posthog/api/authentication.py:
```diff
@@ -178,6 +191,11 @@ def create(self, validated_data: dict[str, str]) -> Any:
+            # Axes tracks failed attempts via authentication signals. If this failure triggered a
+            # lockout, surface the lockout response instead of the generic credential error.
+            if handler.is_locked(axes_request, credentials=axes_credentials):
+                raise AxesBackendPermissionDenied("Account locked: too many login attempts.")
```
Added `test_login_lockout_is_ip_based()` which validates per-IP lockout behavior - complements the existing brute force test and confirms django-ipware 7.0.1 IP detection works correctly.

webjunkie committed:
Merge branch 'master' into chore/axes-upgrade
